toolchain/lex/lex.cpp

```diff
@@ -1322,44 +1425,55 @@ class Lexer::ErrorRecoveryBuffer {
   auto InsertBefore(TokenIndex insert_before, TokenKind kind) -> void {
     CARBON_CHECK(insert_before.index > 0)
         << "Cannot insert before the start of file token.";
+    CARBON_CHECK(insert_before.index <
+                 static_cast<int>(buffer_.token_infos_.size()))
+        << "Cannot insert after the end of file token.";
     CARBON_CHECK(new_tokens_.empty() ||
                  new_tokens_.back().first <= insert_before)
         << "Insertions performed out of order.";
 
+    // If the token we're inserting before had leading whitespace, mark the
+    // inserted token as also having leading whitespace. This avoids changing
+    // whether the prior tokens had leading or trailing whitespace when
+    // inserting.
+    bool insert_leading_space = buffer_.HasLeadingWhitespace(insert_before);
+
     // Find the end of the token before the target token, and add the new token
     // there.
     TokenIndex insert_after(insert_before.index - 1);
     const auto& prev_info = buffer_.GetTokenInfo(insert_after);
     int32_t byte_offset =
-        prev_info.byte_offset + buffer_.GetTokenText(insert_after).size();
+        prev_info.byte_offset() + buffer_.GetTokenText(insert_after).size();
     new_tokens_.push_back(
-        {insert_before,
-         {.kind = kind,
-          .has_trailing_space = buffer_.HasTrailingWhitespace(insert_after),
-          .is_recovery = true,
-          .byte_offset = byte_offset}});
+        {insert_before, TokenInfo(kind, insert_leading_space, byte_offset)});
   }
 
   // Replace the given token with an error token. We do this immediately,
   // because we don't benefit from buffering it.
   auto ReplaceWithError(TokenIndex token) -> void {
     auto& token_info = buffer_.GetTokenInfo(token);
-    token_info.error_length = buffer_.GetTokenText(token).size();
-    token_info.kind = TokenKind::Error;
+    int error_length = buffer_.GetTokenText(token).size();
+    bool has_leading_space = token_info.has_leading_space();
+    int32_t byte_offset = token_info.byte_offset();
```

Comment:
To be clear, I've switched this to a `ResetToError` method. It still does the constructor call in its body so that it gets the payload size checking without duplicating code.

Reply:
SGTM